### PR TITLE
347 state persistence

### PR DESCRIPTION
**Pull request by tafsiri**
@kanitw @vlandham a work-in-progress PR of the persistence API (not ready for merge). I haven't tested the libVoyager functions, but I did test reseting the state from the webapp itself. Take a look and se if the shape of this looks good. I'll also update when I've tested this branch in voyager-electron (after which i'll be adding some test-cases in the repo itself).

As an aside @kanitw, i've added default values to each of the reducers, turns out that redux expects this to be the case. initialState in createStore is actually meant for the use case of restoring for persistence. There is a comment about this from the redux author [here](https://stackoverflow.com/a/36657621)

### CODE REVIEW

**Comment by tafsiri** [comment]
Also it looks like it has some older commits as i started this branch off of my last PR'ed branch. Since those commits have been squashed they still appear here, but the diff is correct.

**Comment by kanitw** [comment]
> Also it looks like it has some older commits as i started this branch off of my last PR'ed branch. Since those commits have been squashed they still appear here, but the diff is correct.

Maybe rebase?

**Comment by kanitw** [comment]
> As an aside @kanitw, i've added default values to each of the reducers, turns out that redux expects this to be the case. initialState in createStore is actually meant for the use case of restoring for persistence. There is a comment about this from the redux author here

Thanks!

**Comment by tafsiri** [comment]
I don't think rebase would work as those commits are not actually in master (due to squash)? Anyway I merged already to catch up. The commits will be squashed out anyway on merge though.

**Comment by kanitw** [comment]
> I don't think rebase would work as those commits are not actually in master (due to squash)? Anyway I merged already to catch up. The commits will be squashed out anyway on merge though.


Oh the trick is to use interactive rebase (`git rebase -i master`) and then drop all the commits that are already squashed to master and keep only the new commits.  :)

**Comment by tafsiri** [comment]
closes #347

**Comment by tafsiri** [comment]
@kanitw @vlandham this is now ready for review. Removing the WIP (work in progress label).

I did see that while the tests pass travis, the main build seems to fail. 

```
64.81s$ npm run build
> datavoyager@2.0.0-alpha.3 build /home/travis/build/vega/voyager
> node scripts/build.js
Creating an optimized production build...
ts-loader: Using typescript@2.3.4 and /home/travis/build/vega/voyager/tsconfig.json
Failed to compile.
(12,27): error TS2307: Cannot find module '../build/src/models/index'.
npm ERR! Linux 4.8.12-040812-generic
npm ERR! argv "/home/travis/.nvm/versions/node/v4.8.3/bin/node" "/home/travis/.nvm/versions/node/v4.8.3/bin/npm" "run" "build"
npm ERR! node v4.8.3
npm ERR! npm  v2.15.11
npm ERR! code ELIFECYCLE
npm ERR! datavoyager@2.0.0-alpha.3 build: `node scripts/build.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the datavoyager@2.0.0-alpha.3 build script 'node scripts/build.js'.
npm ERR! This is most likely a problem with the datavoyager package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node scripts/build.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs datavoyager
npm ERR! Or if that isn't available, you can get their info via:
npm ERR! 
npm ERR!     npm owner ls datavoyager
npm ERR! There is likely additional logging output above.
npm ERR! Please include the following file with any support request:
npm ERR!     /home/travis/build/vega/voyager/npm-debug.log
```

do you have any ideas what could cause this? tests and build run correctly on my local machine.

**Comment by tafsiri** [comment]
never mind my error message. build passed this time. guess it was the lint bugs.

File: src/lib-voyager.tsx [hunk 1]
```diff
@@ -66,6 +66,63 @@ class Voyager {
...
+   * @returns {Function} unsubscribe, call this function to remove this listener.
+   *
+   * @memberof Voyager
+   */
+  public onStateChange(onChange: Function): Function {
+    let currentState: any;
+
+    const handleChange = () => {
+      const nextState = this.store.getState();
+      if (nextState !== currentState) {
```
Comment (kanitw) [inline src/lib-voyager.tsx]: It is quite interesting that you have to check the state one more time here, but I guess that's fine.

**Pull request merged**